### ISSUE TITLE
Long processing time in misc.py for data handling

### ISSUE THREAD

**Issue by eogns282** [iLampard/ep_llm#6]
Hi,

Thank you for sharing the LAMP project. I've been experimenting with your code, particularly interested in prediction tasks on the GDELT data.

However, I've encountered a significant performance issue that I hope to get some guidance on. For "Step 3: we setup the samples to train the ranking model", when executing the script referenced in [this line of code](https://github.com/iLampard/lamp/blob/c48312108ef9b0eb54c9e4581b2af248abc09094/utils/misc.py#L479), the processing time seems excessively long for the amount of data being handled. A 24-hour run did not complete processing even 50 data items, which seems unusual. 

I'm running this on a server equipped with GPU 3090s, and both CPU and memory specs are sufficiently high, so I believe hardware limitations are unlikely to be the cause.

Could there be any specific configuration or optimization I might be missing, or is this processing time expected for the task at hand?

Thank you for your time and assistance.

**Comment by iLampard** [comment]
Hi, 

If i remember right, the event retrieval should not be taking too long time. what type of distance are you using?  edit distance?

**Comment by eogns282** [comment]
Dooray! Failure Notice



  
  
    
    
  
  
    
        Failure Notice
    
  
  
    
        
        Your message sent to
          
***@***.***) 
          has failed to be delivered.
        
        
          Please refer to the below for details.
        
    
  
  
    
      
        
          
            * Recipient : 
***@***.***) 
          
          
            * Sent time : 
2024-04-04T01:08:24
          
          
            * Subject : 
Re: [iLampard/lamp] Long processing time in misc.py for data handling (Issue #6)
          
          
            
              
                * Remote host said : 
              
              
                
Your mail was denied from the receiver.
                
              
            
                      
          
        
      
    
  
  
    
        
          This message was sent from a notification-only address that cannot accept incoming email.
          For more information, please contact           ***@***.***
        
    
  
  
    
        © Dooray!.

**Comment by eogns282** [comment]
I am using the default distance metric specified in the code, which is 'bert'. I haven't made any changes to that configuration.

Additionally, I realized that I'm using a newer version of PyTorch than what is listed in the README. Could this be the reason for the long processing times I'm experiencing? 

Thank you again for your assistance.

**Comment by iLampard** [comment]
Hi

I will have a look at this and get back to you shortly

**Comment by iLampard** [comment]
Hi, 

I have investigated the function retrieve_event_from_source_data and test it again on my m1 mac. 

setting: torch==2.1.2


running script:
``` 
  
    source_data_fn = 'gdelt_emb.pkl'
    pred_data_fn = '../../logs/ke_gdelt_test.pkl'
    source_data = file_uri_reader_processor(source_data_fn)['data']
    pred_data = file_uri_reader_processor(pred_data_fn)

    make_samples_for_energy_function(
        gpt_db_name='gdelt_chatgpt/object',
        source_data=source_data,
        pred_data=pred_data,
        pred_type='relation',
        topk=5,
        # pred_type='object',
        # topk=20,
        ebm_db_name='ke_gdelt_bert_ebm_dataset_v0509'
        retro_top_n=2,
        distance_type='bert')
``` 


i put a timer inside the function **retrieve_event_from_source_data**

```         
elif distance_type == 'bert':

    target_text = encoder_model.encode(target_event)

    start_time = time.perf_counter()

    scores = [event_distance(target_text, event[-1], distance_type=distance_type) for event in source_data_]

    end_time = time.perf_counter()

    elapsed_time = end_time - start_time

    print(f"The code took {elapsed_time} seconds to run.")
``` 

the score constantly takes 2-3 secs to run. The source_data_ contains ~10k event vectors(already computed) and then we compute the distance between the query event and all these 10k events, which normally should not take a very long time. 

Can you add a timer in your code to see which submodule works abnormally long?
